### PR TITLE
865-DataSubset-split-to-Load/Process/Layer

### PR DESCRIPTION
**Pull request by Ushcode**
Addressing issue #865, this PR implements a refactor of  `PostgresDataSubSet` class as well ats its children to allow a `dataSubset` to be specified without a subdirectory. This means that SQL post processing and GeoServer uploading can happen independently of data upload and arbitrary action can be performed on prior uploaded data.

`TBoxCSV` and `RDF` data are still required to specify a subdirectory, while `PostGresDataSubset` and its children: `Tabular`, `GeoServerDataSubset`, `Vector` and `Raster` will all proceed with a warning that no subdirectory is specified.

Correct behaviour of these changes was tested and output logs of the uploads are attached to this PR

# RDF error

15:31:40.820 [main] DEBUG org.eclipse.jetty.util.component.AbstractLifeCycle - STOPPED com.bigdata.rdf.sail.webapp.client.AutoCloseHttpClient@3ef41c66
Exception in thread "main" java.lang.RuntimeException: No Subdirectory Specified - required for RDF data
        at com.cmclinnovations.stack.clients.core.datasets.RDF.loadInternal(RDF.java:18)
        at com.cmclinnovations.stack.clients.core.datasets.DataSubset.load(DataSubset.java:55)
        at com.cmclinnovations.stack.clients.core.datasets.DatasetLoader.lambda$loadData$6(DatasetLoader.java:142)
        at java.base/java.util.ArrayList.forEach(Unknown Source)
        at com.cmclinnovations.stack.clients.core.datasets.DatasetLoader.loadData(DatasetLoader.java:142)
        at java.base/java.util.Iterator.forEachRemaining(Unknown Source)
        at com.cmclinnovations.stack.clients.core.datasets.DatasetLoader.uploadInputDatasets(DatasetLoader.java:61)
        at com.cmclinnovations.stack.DataUploader.main(DataUploader.java:12)
 
# TBoxCSV error

15:33:34.413 [main] DEBUG org.eclipse.jetty.util.component.AbstractLifeCycle - STOPPED SslContextFactory@58740366(null,null)
15:33:34.413 [main] DEBUG org.eclipse.jetty.util.component.AbstractLifeCycle - STOPPED com.bigdata.rdf.sail.webapp.client.AutoCloseHttpClient@30ca0779
Exception in thread "main" java.lang.RuntimeException: No subdirectory specified - required for TBoxCSV data
        at com.cmclinnovations.stack.clients.core.datasets.TBoxCSV.loadInternal(TBoxCSV.java:32)
        at com.cmclinnovations.stack.clients.core.datasets.DataSubset.load(DataSubset.java:55)
        at com.cmclinnovations.stack.clients.core.datasets.DatasetLoader.lambda$loadData$6(DatasetLoader.java:142)
        at java.base/java.util.ArrayList.forEach(Unknown Source)
        at com.cmclinnovations.stack.clients.core.datasets.DatasetLoader.loadData(DatasetLoader.java:142)
        at java.base/java.util.Iterator.forEachRemaining(Unknown Source)
        at com.cmclinnovations.stack.clients.core.datasets.DatasetLoader.uploadInputDatasets(DatasetLoader.java:61)
        at com.cmclinnovations.stack.DataUploader.main(DataUploader.java:12)

### CODE REVIEW

**Review by Ushcode — COMMENTED** [review]
# Routing stack

Everything uploads and processes correctly, following warning is present in console:

```15:55:05.445 [main] WARN com.cmclinnovations.stack.clients.core.datasets.PostgresDataSubset - No Subdirectory specified, Continuing with SQL process and creation without data upload```

File: Deploy/stacks/dynamic/stack-data-uploader/README.md [hunk 1]
```diff
@@ -62,15 +63,16 @@ By default all dataset configuration files in the [`inputs/configs/`](./inputs/c
...
     dataset1.json       # Configuration file for dataset1
     dataset2.json       # Configuration file for dataset2
   data/                 # Directory in which the data files should be stored
     dataset1/           # Data directory for dataset1
-      data.csv          # Only one data subset so no need for a subdirectory
+      subdirectory/     # Data subdirectory for dataset1
+        data.csv        # Only one data subset so no need for a subdirectory
     dataset2/           # Data directory for dataset2
       datasubset1/      # Data directory for data subset1
         polygon.geojson # Data file
```
Comment (gpeb2) [inline Deploy/stacks/dynamic/stack-data-uploader/README.md:78]:
```suggestion
      datasubset1/     # Data subdirectory for data subset1
        data.csv        # Data file for dataset1
    dataset2/           # Data directory for dataset2
      datasubset2/      # Data subdirectory for data subset2
        polygon.geojson # Data file for dataset2
```

File: Deploy/stacks/dynamic/stack-clients/src/main/java/com/cmclinnovations/stack/clients/core/datasets/PostgresDataSubset.java [hunk 2]
```diff
@@ -37,9 +42,13 @@ public boolean usesPostGIS() {
     @Override
     void loadInternal(Dataset parent) {
         String database = parent.getDatabase();
-        Path dataSubsetDirectory = parent.getDirectory().resolve(this.getSubdirectory());
-        PostGISClient.getInstance().resetSchema(database);
-        loadData(dataSubsetDirectory, database, parent.baseIRI());
+        Path subdirectory = this.getSubdirectory();
+        if (null != subdirectory) {
+            Path dataSubsetDirectory = parent.getDirectory().resolve(subdirectory);
+            loadData(dataSubsetDirectory, database, database);
```
Comment (gpeb2) [inline Deploy/stacks/dynamic/stack-clients/src/main/java/com/cmclinnovations/stack/clients/core/datasets/PostgresDataSubset.java:48]:
```suggestion
            loadData(dataSubsetDirectory, database, parent.baseIRI());
```

File: Deploy/stacks/dynamic/stack-clients/src/main/java/com/cmclinnovations/stack/clients/core/datasets/RDF.java [hunk 3]
```diff
@@ -11,8 +13,12 @@ public boolean usesBlazegraph() {
 
     @Override
     void loadInternal(Dataset parent) {
+        Path subdirectory = this.getSubdirectory();
+        if (null == subdirectory) {
+            throw new RuntimeException("No Subdirectory Specified - required for RDF data");
```
Comment (gpeb2) [inline Deploy/stacks/dynamic/stack-clients/src/main/java/com/cmclinnovations/stack/clients/core/datasets/RDF.java:18]:
```suggestion
            throw new RuntimeException("No 'subdirectory' specified - required for RDF data");
```

File: Deploy/stacks/dynamic/stack-clients/src/main/java/com/cmclinnovations/stack/clients/core/datasets/TBoxCSV.java [hunk 4]
```diff
@@ -27,7 +27,11 @@ void loadInternal(Dataset dataset) {
         RemoteStoreClient remoteStoreClient = BlazegraphClient.getInstance()
                 .getRemoteStoreClient(dataset.getNamespace());
 
-        Path datasubsetDir = dataset.getDirectory().resolve(getSubdirectory());
+        Path subdirectory = this.getSubdirectory();
+        if (null == subdirectory) {
+            throw new RuntimeException("No subdirectory specified - required for TBoxCSV data");
```
Comment (gpeb2) [inline Deploy/stacks/dynamic/stack-clients/src/main/java/com/cmclinnovations/stack/clients/core/datasets/TBoxCSV.java:32]:
```suggestion
            throw new RuntimeException("No 'subdirectory' specified - required for TBoxCSV data");
```

**Review by gpeb2 — CHANGES_REQUESTED** [review]
I've made some some suggestions, happy to discuss them.

**Review by gpeb2 — APPROVED** [review]

**Pull request merged**